### ISSUE TITLE
LGTM.com - false positive This assignment to is useless, since its value is never read.

### ISSUE THREAD

**Issue by jmarolf** [github/codeql#6785]
**Description of the false positive**

In the code below it warns that `spanDirective` is never read even though the value is passed to a method in which it is read:

```csharp
if (directiveNode is LineSpanDirectiveTriviaSyntax spanDirective)
{
    return GetLineSpanDirectiveEntry(spanDirective, directiveLineNumber);
}
```

The compiler transforms the code into this (you can view the IL [here](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFDKgZgASwwGQJYB2ApgMoAOAhgQCJ4BORAxgC54BuRAKne3hSQE8CzCgA8MAb2QYZWTIWa5CRDAF4AfBgAsMANzIAvslSZsAYUnTZ6DAowBZABQB5AEYArJouD0v7IgByAPbAKgCUlkiy0bYAZhiOPgws/sGhtgDOSsTkVLTJrBzcvPxCIuIZlDS+KRxhVjGRjY1QAOwYAOJEzPg5Vfl+HACiwnQCjpV5NYVEYfpRzUYLjQ0xbRihsRQArgA2zPPRS9GrZDxsFMwqdl09yrnVBf4jzGOOvaT90/7FbHyCwjEGEmj0Gs1WUmWa3aIIGtSIADoPodZEsDEA===))
```csharp
LineSpanDirectiveTriviaSyntax lineSpanDirectiveTriviaSyntax = directiveNode as LineSpanDirectiveTriviaSyntax;
if (lineSpanDirectiveTriviaSyntax != null)
{
    return GetLineSpanDirectiveEntry(spanDirective, directiveLineNumber);
}
```

So this is a false positive unless we can prove that there exists no conversion to `LineSpanDirectiveTriviaSyntax` that can ever succeed.

**URL to the alert on the project page on LGTM.com**

https://lgtm.com/projects/g/dotnet/roslyn/snapshot/c9e0dd1ec0e440897388bbd55184fb53a555e3a4/files/src/Compilers/CSharp/Portable/Syntax/CSharpLineDirectiveMap.cs?sort=name&dir=ASC&mode=heatmap#xfcee166e229ccb35:1

**Comment by MathiasVP** [comment]
Indeed, this looks like a false positive. Thank you for reporting it!

I'll forward this to the C# team. Our current focus is on improving our security analysis, and since this isn't a security query. we'll put this on our backlog and prioritize it if we get enough reports of the same underlying issue in other projects. If you think that your report is related to our security analysis, please clarify that in a comment. Either way, we'll let you know here as soon as it's fixed!

As you might know, all of our queries are open source. If you do have an idea for a code change, we encourage you to open a pull request. Both GitHub Code Scanning and lgtm.com have facilities for suppressing individual alerts or disabling a query.

**Comment by hvitved** [comment]
This happens because `dotnet/roslyn` uses .NET6 preview, which is not supported by the C# CodeQL extractor (the extractor is based on [Microsoft.CodeAnalysis](https://www.nuget.org/packages/Microsoft.CodeAnalysis) 3.9.0). The file in question (`CSharpLineDirectiveMap.cs`) has compilation errors (see [here](https://lgtm.com/query/1811050089472871562/)), and therefore results in that file cannot be trusted.

**Comment by GwynethLlewelyn** [comment]
> This happens because `dotnet/roslyn` uses .NET6 preview, which is not supported by the C# CodeQL extractor

Hmm. Ok, _that_ explains why I can't even get to _start_ the analysis at all — I've gone full .NET6. Still, in the meantime, Microsoft.CodeAnalysis has been bumped to 4.0.1 (with a prerelease 4.1.0 just out), so I wonder if CodeQL will start accepting .NET6-only builds soon-ish?...

**Comment by hvitved** [comment]
@GwynethLlewelyn : Sorry for the late reply. We are working on supporting .NET 6 / C# 10, and we are hoping to support it within a couple of months.

**Comment by GwynethLlewelyn** [comment]
@hvitved no apologies needed, after all, you're doing so much work on our behalf... anyway, that sounds exciting! Thanks so much for keeping us posted! I'm looking forward to test it out.

Happy coding :)
